packages/compiler/core/parser.ts

```diff
@@ -1060,6 +1133,617 @@ export function parse(code: string | SourceFile, options: ParseOptions = {}): Ca
+      id,
+      decorators,
+      value,
+      optional,
+      default: defaultValue,
+      ...finishNode(pos),
+    };
+  }
+
+  function parseProjectionIfExpression(): ProjectionIfExpressionNode {
+    const pos = tokenPos();
+    parseExpected(Token.IfKeyword);
+    const test = parseProjectionExpression();
+    const consequent = parseProjectionBlockExpression();
+    let alternate = undefined;
+    if (parseOptional(Token.ElseKeyword)) {
+      if (token() === Token.IfKeyword) {
+        alternate = parseProjectionIfExpression();
+      } else {
+        alternate = parseProjectionBlockExpression();
+      }
+    }
+
+    return {
+      kind: SyntaxKind.ProjectionIfExpression,
+      test,
+      consequent,
+      alternate,
+      ...finishNode(pos),
+    };
+  }
+
+  function parseProjectionBlockExpression(): ProjectionBlockExpressionNode {
+    const pos = tokenPos();
+    parseExpected(Token.OpenBrace);
+    const statements = parseProjectionStatementList();
+    parseExpected(Token.CloseBrace);
+    return {
+      kind: SyntaxKind.ProjectionBlockExpression,
+      statements,
+      ...finishNode(pos),
+    };
+  }
+
+  function parseProjectionTupleExpression(): ProjectionTupleExpressionNode {
+    const pos = tokenPos();
+    const values = parseList(ListKind.Tuple, parseProjectionExpression);
+    return {
+      kind: SyntaxKind.ProjectionTupleExpression,
+      values,
+      ...finishNode(pos),
+    };
+  }
+  function parseProjectionSelector():
+    | IdentifierNode
+    | MemberExpressionNode
+    | ProjectionInterfaceSelectorNode
+    | ProjectionModelSelectorNode
+    | ProjectionOperationSelectorNode
+    | ProjectionUnionSelectorNode
+    | ProjectionEnumSelectorNode {
+    const pos = tokenPos();
+    const selectorTok = expectTokenIsOneOf(
+      Token.Identifier,
+      Token.ModelKeyword,
+      Token.OpKeyword,
+      Token.InterfaceKeyword,
+      Token.UnionKeyword,
+      Token.EnumKeyword
+    );
+
+    switch (selectorTok) {
+      case Token.Identifier:
+        return parseIdentifierOrMemberExpression();
+      case Token.ModelKeyword:
+        nextToken();
+        return {
+          kind: SyntaxKind.ProjectionModelSelector,
+          ...finishNode(pos),
+        };
+      case Token.OpKeyword:
+        nextToken();
+        return {
+          kind: SyntaxKind.ProjectionOperationSelector,
+          ...finishNode(pos),
+        };
+      case Token.InterfaceKeyword:
+        nextToken();
+        return {
+          kind: SyntaxKind.ProjectionInterfaceSelector,
+          ...finishNode(pos),
+        };
+      case Token.UnionKeyword:
+        nextToken();
+        return {
+          kind: SyntaxKind.ProjectionUnionSelector,
+          ...finishNode(pos),
+        };
+      case Token.EnumKeyword:
+        nextToken();
+        return {
+          kind: SyntaxKind.ProjectionEnumSelector,
+          ...finishNode(pos),
+        };
+    }
+
+    throw new Error("Unknown token kind");
```

Comment:
I think I need to do something for recovery here but I'm not sure what (like, what happens if you put `(` or `+` in there). I think this should be fine while typing most selectors since you'll get an id on the way up to the keyword but this error should be doing something smart. /cc @nguerrera

Reply:
Could you provide a full snippet where this would currently throw here? I think that will help me to think about the recovery options.

Reply:
I think this will throw -

```
projection +
```

Reply:
Oh, I see. I think you can insert a "missing identifier" as a reference expression here. return ParseReferenceExpression("projectionSelector") then add `"projectionSelector": "<Message of your choice>"` to "token-expected" in messages.ts.

Reply:
This is vital to fix. You hit it as soon as you type "projection <nothing yet>" -> language server crashes.

Reply:
Need to fix both the cause and the effect here.  The language server crashing for any reson relating to evaluating a single spec is catastrophic for the user experience.

Reply:
We can make the language server not shut down when compile fails with exception, if that's what you mean for "and the effect" here, but that would be separate from this change.

Reply:
(Side rant: We also need to be careful about designing experiences for bug cases. I have seen my fair share of software that does this in the name of "robustness" let bugs sit for years whereas crashes get fixed promptly. So while we can continue, we need it to be visible enough that something terrible has happened and needs to be fixed.)

Reply:
OK, as of #94, the language server will no longer tear down no matter how the compilation fails with an exception. *But* we will log an "internal compiler error, file bug, stack trace" like on the command line in this case. 

(Rant continued: It is *really* important to me that we don't go down the route of blaming users for bugs in the compiler. I cry inside when programs are like "You did something wrong: object instance not set to an instance of an object." These programs hold on to their bugs longer. I sometimes compare this with my experience teaching some folks to ski. Some would develop sub-optimal technique but be proud that they "didn't fall". While others who were less afraid to fall, just picked themselves up each time and learned to ski better faster.)

Anything we expect to happen outside our control should still not trigger this last resort handler. So at a minimum for this PR, we should at least have a better catch somewhere at the top of where we can hit these exceptions and log a diagnostic that indicates that we do expect this to happen for reasons related to the input code and not because there's a bug. And this handler should distinguish between these expected cases and an arbitrary bug, perhaps by having a well-known exception type. 

And longer term, we should think about improving the quality of those diagnostics so that users can locate the source of the problem better, etc. I also think it would be preferable if there was error recovery to treat the failed projected type behaved somewhat like an unresolved type reference, and continued to analyze for other errors.

Reply:
The "so at a minimum" above was meant for https://github.com/microsoft/cadl/pull/16#discussion_r763290674 not this line of code. Here we should do https://github.com/microsoft/cadl/pull/16#discussion_r754594892